### ISSUE TITLE
Lazy loading results in IndexOutOfBoundsException 

### ISSUE THREAD

**Issue by ThijsMaenhout** [vaadin/vaadin-combo-box-flow#154]
Hi,

When populating  a combobox with data from a pageable backend I get the following exception:

java.lang.IndexOutOfBoundsException: Index: 150, Size: 150
	at java.util.ArrayList.rangeCheck(ArrayList.java:657) ~[na:1.8.0_152]
	at java.util.ArrayList.get(ArrayList.java:433) ~[na:1.8.0_152]
	at com.vaadin.flow.data.provider.DataCommunicator.lambda$getJsonItems$3(DataCommunicator.java:597) ~[flow-data-1.2.0.jar:1.2.0]
	at java.util.stream.IntPipeline$4$1.accept(IntPipeline.java:250) ~[na:1.8.0_152]
	at java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:110) ~[na:1.8.0_152]
	at java.util.Spliterator$OfInt.forEachRemaining(Spliterator.java:693) ~[na:1.8.0_152]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[na:1.8.0_152]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[na:1.8.0_152]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[na:1.8.0_152]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:1.8.0_152]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[na:1.8.0_152]
	at com.vaadin.flow.data.provider.DataCommunicator.getJsonItems(DataCommunicator.java:599) ~[flow-data-1.2.0.jar:1.2.0]
	at com.vaadin.flow.data.provider.DataCommunicator.lambda$collectChangesToSend$2(DataCommunicator.java:562) ~[flow-data-1.2.0.jar:1.2.0]
	at com.vaadin.flow.data.provider.DataCommunicator.applyIfNotEmpty(DataCommunicator.java:613) ~[flow-data-1.2.0.jar:1.2.0]
	at com.vaadin.flow.data.provider.DataCommunicator.withMissing(DataCommunicator.java:607) ~[flow-data-1.2.0.jar:1.2.0]
	at com.vaadin.flow.data.provider.DataCommunicator.collectChangesToSend(DataCommunicator.java:561) ~[flow-data-1.2.0.jar:1.2.0]
	at com.vaadin.flow.data.provider.DataCommunicator.flush(DataCommunicator.java:464) ~[flow-data-1.2.0.jar:1.2.0]
	at com.vaadin.flow.data.provider.DataCommunicator.lambda$requestFlush$2f364bb9$1(DataCommunicator.java:421) ~[flow-data-1.2.0.jar:1.2.0]
	at com.vaadin.flow.internal.StateTree.lambda$runExecutionsBeforeClientResponse$1(StateTree.java:350) ~[flow-server-1.1.0.jar:1.1.0]
	at java.util.ArrayList.forEach(ArrayList.java:1257) ~[na:1.8.0_152]
	at com.vaadin.flow.internal.StateTree.runExecutionsBeforeClientResponse(StateTree.java:347) ~[flow-server-1.1.0.jar:1.1.0]
	at com.vaadin.flow.server.communication.UidlWriter.encodeChanges(UidlWriter.java:334) ~[flow-server-1.1.0.jar:1.1.0]
	at com.vaadin.flow.server.communication.UidlWriter.createUidl(UidlWriter.java:179) ~[flow-server-1.1.0.jar:1.1.0]
	at com.vaadin.flow.server.communication.UidlRequestHandler.writeUidl(UidlRequestHandler.java:119) ~[flow-server-1.1.0.jar:1.1.0]
	at com.vaadin.flow.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:91) ~[flow-server-1.1.0.jar:1.1.0]
	at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40) ~[flow-server-1.1.0.jar:1.1.0]
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1533) ~[flow-server-1.1.0.jar:1.1.0]
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:300) [flow-server-1.1.0.jar:1.1.0]
	at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:61) [vaadin-spring-10.1.0.jar:na]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) [javax.servlet-api-3.1.0.jar:3.1.0],

I'm able to populate the combobox with 50 items (default page size), It is possible to filter, but when I scroll in the combobox to the next range (eg: 50 - 100).  I get this exception. 

I'm working on 12.0.0.BETA_1.

**Comment by ThijsMaenhout** [comment]
Hi,

I did create a starterproject and rebased it to the 12.0.0.BETA1 Branch.

I adjusted the code so it creates 1500 products.  When adding a new order and when you open the products combobox and scroll down, you can replicate the bug.

code can be found here:

https://github.com/ThijsMaenhout/Vaadin-TestCombobox.git

**Comment by ThijsMaenhout** [comment]
I just tested this issue with the spring Starter project (https://vaadin.com/start/pre-release) and I have the same issue.  It was tested with 12.0.0.BETA2

Add this code in the datagenerator (line 300) : 
   for (int i = 0; i < 1000 ; i++) {
            Product product = new Product();
            product.setName(java.util.UUID.randomUUID().toString());
            double doublePrice = 2.0 + random.nextDouble() * 100.0;
            product.setPrice((int) (doublePrice * 100.0));
            products.add(productsRepo.save(product));
        }

**Comment by ThijsMaenhout** [comment]
Any suggestions on how I can solve / investigate this?
Thank you

**Comment by pekam** [comment]
Hi, @ThijsMaenhout!

Could you provide a more isolated use-case for the bug? For example, start with the Project Base and add only the minimal amount of code to reproduce the issue. Based on the full-stack examples it's hard to say if the problem is in the application code or in the component.

One thing that you could check is that does the `DataProvider.size` method return the correct number of items.

**Comment by ThijsMaenhout** [comment]
I was able to solve this by using a different DataProvider.  By using PageableDataProvider , i managed to get it working.  

Perhaps it would be good to check if the DataProvider is Pageable or not when performing Lazy Loading.

**Comment by aibar** [comment]
Same issue in vaadin 13.0.4

Steps to reproduce:
1.  Download the latest bakery flow starter
2. Change Datagenerator class to produce 300 products
3. Open Storefront tab
4. New Order
5. Open products combobox
6. Try to scroll -> BOOM

Extending from PageableDataProvider from spring-data-provider did not resolved the issue.

**Comment by andytudor** [comment]
Same issue in version 14.0.0.rc6

**Comment by pekam** [comment]
The problem seems to be in Bakery/dataproviders, or in the way that the items are added.

I added 100 items in the same way as in [this comment](https://github.com/vaadin/vaadin-combo-box-flow/issues/154#issuecomment-439314665) and had a look at the responses that `ProductDataProvider` provides to `ComboBox`'s queries:
- `sizeInBackend`: returns 212, which seems incorrect (but anyway it says that there should be items to fetch for the second page, i.e. after the first 50 items)
- `fetchFromBackend` returns 0 items for the second page query

So the size-callback says that there should be enough items to fetch for this query, but then no items are actually returned. It's expected that `ComboBox` will break in this case.

But I don't know if that is a proper way of adding items to the backend in Bakery (looks a bit hacky), or if it should be handled some other way to make everything work properly. Ping @alvarezguille.

**Comment by DiegoCardoso** [comment]
As pointed out by @pekam, the issue was on the Bakery side, in the way it was retrieving the data.

`ComboBox` asks for data giving the offset and limit, while Spring's `PageRequest` expects a page and size.

To solve this, we changed the implementation to use [FilterablePageableDataProvider](https://vaadin.com/directory/component/spring-data-provider-add-on) to avoid issues between Pageable and Query.

A new version with that fix was released and can be downloaded at [Vaadin Bakery App Starter for Flow and Spring](https://vaadin.com/start/latest/full-stack-spring).
